forum/qa-plugin/user-changeable-theme/user-theme-layer.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+class qa_html_theme_layer extends qa_html_theme_base
+{
+    private $cookie = 'qa_user_theme';
+
+    private function check_theme()
+    {
+        if (isset($_COOKIE[$this->cookie])) {
+            return (bool)$_COOKIE[$this->cookie];
+        } elseif ($user = qa_get_logged_in_userid()) {
+            $sql = 'SELECT `theme` FROM `^users` WHERE `userid` = #';
+            $theme = qa_db_read_one_value(qa_db_query_sub($sql, $user));
+            setcookie($this->cookie, $theme, time()+31556926, '/', QA_COOKIE_DOMAIN);
+            return (bool)$theme;
+        }
+        return false;
+    }
+
+    private function save_theme($theme)
+    {
+        if ($theme == 0 || $theme == 1) {
+            if ($this->check_theme() !== (bool)$theme) {
+                if ($user = qa_get_logged_in_userid()) {
+                    $sql = 'UPDATE `^users` SET `theme` = # WHERE `userid` = #';
+                    qa_db_query_sub($sql, $theme, $user);
+                }
+                setcookie($this->cookie, $theme, time()+31556926, '/', QA_COOKIE_DOMAIN);
+            }
+        }
+        qa_redirect(qa_request());
+    }
+
+    public function head_lines()
+    {
+        if (qa_opt('user_theme_enable')) {
+            if (isset($_POST['select_theme'])) {
+                $this->save_theme($_POST['select_theme']);
+            }
+            $this->content['head_lines'][] = '<style>
```

Comment:
Byłoby najlepiej, ale chcieliśmy uniknąć dołączania aż tylu plików, tym bardziej że tutaj kodu jest niedużo. Najprościej byłoby dorzucić to do głównego pliku z CSSem SnowFlat, ale też chcieliśmy, aby plugin można byłoby swobodnie w całości wyciągnąć.